Do not process a range from rangeGCQueue when it has not yet been initialized

kkaneda opened the pull request:
This might happen only in tests, but the following test caused a panic if the range GC queue tries to process the bootstrap range while `replicateRange()` is called.

```
func TestRangeGCQueue(t *testing.T) {
     defer leaktest.AfterTest(t)
     mtc := &multiTestContext{}
     mtc.Start(t, 3)
     defer mtc.Stop()
     raftID := int64(1)
     mtc.replicateRange(raftID, 0, 1, 2)
}
```

The panic message is following:

```
L="FATAL" T="5/16 01:24:02.783" F="range.go:442" Msg="own replica missing in range at store 3"
```

Here is what happened (I think):

1) `replicateRange()` calls `ChangeReplicas()`, which starts a txn to update the range descriptor.
2) Store 3 receives a Raft message from an unknown Raft group (ID 1).
3) Store 3 creates a new range with Raft ID 1 via `GroupStorage()`. The method does not populate a replica list.
4) Before the txn is committed and the range's replica list is updated, the range GC queue tries to process the range.
5) `GetReplica()` causes a panic since no replica is found.

Alternative solutions:
- Make `GroupStorage()` somehow populate a replica list.
- Disable the scanner in `multiTestContext`.

kkaneda committed:
Do not process a range from rangeGCQueue when it has not yet been initialized

This might happen only in tests, but the following test caused a
panic if the range GC queue tries to process the bootstrap range
while replicateRange() is called.

func TestRangeGCQueue(t *testing.T) {
     defer leaktest.AfterTest(t)
     mtc := &multiTestContext{}
     mtc.Start(t, 3)
     defer mtc.Stop()
     raftID := int64(1)
     mtc.replicateRange(raftID, 0, 1, 2)
}

The panic message is following:

L="FATAL" T="5/16 01:24:02.783" F="range.go:442" Msg="own replica missing in range at store 3"

Here is what happened (I think):

1) replicateRange() calls ChangeReplicas(), which starts a txn to update the range descriptor.
2) Store 3 receives a Raft message from an unknown Raft group (ID 1).
3) Store 3 creates a new range with Raft ID 1. GroupStorage() does not populate a replica list.
4) Before the range's replica list is updated, the range GC queue tries to process the range.
5) GetReplica() causes a panic since no replica is found.

Checking isInitialized addresses the failure.

tschottdorf commented:
It seems somehow reasonable that `GroupStorage()` shouldn't pass out an empty replica list given that it knows that the requested range should be in it (?). I'm not sure about the implications though, and it might not matter if the naive solution is the correct one: the range gc queue should check `r.isInitialized()` (which is only true after a snapshot has been applied, which in turn updates the descriptor accordingly). That of course raises the question of when ranges which were added at some point but then never received a snapshot are garbage collected.

It seems awkward to have `GetReplica` and `GetReplicaOptional`. Either it should return an error in general or panic, depending on how we go about the above. From the way the GC queue uses it, an error seems appropriate.

kkaneda commented:
Thanks for the review, @tschottdorf . Adding `r.isInitialized()` seems to work (at least no panic happens after multiple runs)! Updated the PR. Not garbage-collecting ranges might be fine if it won't happen very frequently? Not sure..

`GetReplica` is called also from `kv/local_sender.`. Returning an error might be appropriate as well for this case, but I wasn't totally sure. Just removed `GetReplicaOptional` and kept `GetReplica` as-is.

Unrelated to this PR, but I have one question.. GroupStorage()` calls `addRangeInternal`, which looks at the range's start key to sort `rangesByKey` after insertion. So, is it correct that a range might not be added to a right place in `rangesByKey` since the range's start key is empty until a snapshot is applied? If so, is there any code that resort ranges later (maybe when another new range is added)? Just wanted to make sure converting  `rangesByKey` to btree won't break anything.

Thanks!

bdarnell commented:
We should probably put a check for IsInitialized in baseQueue; none of the queues can really handle uninitialized ranges right now. (Ironically, the only thing that does make sense for uninitialized ranges is GC, but I think we might need to do that with something separate).

I have a separate change to remove the log.Fatal from GetReplica and just make it return nil instead.

It sounds like a bug that the range is added to rangesByKey while its start key is empty; we need to fix that and add it to the list once it becomes initialized.

kkaneda commented:
Ah, thanks, @bdarnell . I see. So it seems adding a range to `rangesByKey` after it is initialized like a better fix. The scanner won't see any uninitialized ranges. Will file a ticket to track this.

I can still add a check to `baseQueue`, but it looks unnecessary once we make the above change. We also don't need a quick fix as we have #1051. Will closing the PR.
